[HLSL] Support vector swizzles on scalars

llvm-beanz opened the pull request:
HLSL supports vector swizzles on scalars by implicitly converting the scalar to a single-element vector. This syntax is a convienent way to initialize vectors based on filling a scalar value.

There are two parts of this change. The first part in the Lexer splits numeric constant tokens when a `.x` or `.r` suffix is encountered. This splitting is a bit hacky but allows the numeric constant to be parsed separately from the vector element expression. There is an ambiguity here with the `r` suffix used by fixed point types, however fixed point types aren't supported in HLSL so this should not cause any exposable problems (a separate issue has been filed to track validating language options for HLSL: #67689).

The second part of this change is in Sema::LookupMemberExpr. For HLSL, if the base type is a scalar, we implicit cast the scalar to a one-element vector then call back to perform the vector lookup.

Fixes #56658 and #67511

llvm-beanz committed:
[HLSL] Support vector swizzles on scalars

HLSL supports vector swizzles on scalars by implicitly converting the
scalar to a single-element vector. This syntax is a convienent way to
initialize vectors based on filling a scalar value.

There are two parts of this change. The first part in the Lexer splits
numeric constant tokens when a `.x` or `.r` suffix is encountered. This
splitting is a bit hacky but allows the numeric constant to be parsed
separately from the vector element expression. There is an ambiguity
here with the `r` suffix used by fixed point types, however fixed point
types aren't supported in HLSL so this should not cause any exposable
problems (a separate issue has been filed to track validating language
options for HLSL: #67689).

The second part of this change is in Sema::LookupMemberExpr. For HLSL,
if the base type is a scalar, we implicit cast the scalar to a
one-element vector then call back to perform the vector lookup.

Fixes #56658 and #67511

cor3ntin commented on clang/lib/Lex/Lexer.cpp:
```diff
@@ -1950,6 +1950,10 @@ bool Lexer::LexNumericConstant(Token &Result, const char *CurPtr) {
     CurPtr = ConsumeChar(CurPtr, Size, Result);
     PrevCh = C;
+    if (LangOpts.HLSL && C == '.' && (*CurPtr == 'x' || *CurPtr == 'r')) {
+      CurPtr--;
```
Indeed!

AaronBallman commented on clang/test/SemaHLSL/Types/BuiltinVector/ScalarSwizzleErrors.hlsl:
```diff
@@ -0,0 +1,14 @@
+}
+
+int4 SomeNonsense(int V) {
+  return V.poop; // expected-error{{illegal vector component name 'p'}}
```
I'd like another test that validates we catch multiple levels of dots. e.g.,
```
float2 HowManyFloats(float V) {
  return V.rr.rr;
}
```
or dots not followed by anything useful:
```
float2 WhatIsHappening(float V) {
  return V.;
}
```

llvm-beanz replied:
The really sad thing is that the first case there seems to be valid on both our reference compilers:

[DXC on Compiler Explorer](https://godbolt.org/z/KETGPbdfq)
[FXC on Shader Playground](https://shader-playground.timjones.io/e91ae2cfb95892f3676a71fcb2920756)

We may need to support that. I'll add tests either way.

AaronBallman replied:
Please also add CodeGen tests that show we do... whatever *that* is.

AaronBallman commented on clang/test/SemaHLSL/Types/BuiltinVector/ScalarSwizzles.hlsl:
```diff
@@ -0,0 +1,78 @@
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'unsigned int __attribute__((ext_vector_type(1)))' <VectorSplat>
+// CHECK-NEXT: IntegerLiteral {{.*}} 'unsigned int' 1
+
+uint3 FillOneUnsigned(){
+  return 1u.xxx;
+}
```
What happens with:
```
auto HooBoy() {
  return 4wb.xxxx;
}
```
or with a float that has a trailing period followed by this Very Special™ suffix?
```
float3 err() {
  return 1..rrr;
}
```

llvm-beanz replied:
`HooBoy` fails because `wb` isn't a supported suffix in C++ mode (and all HLSL versions are C++-based). Also, even the most experimental HLSL version is only C++-11 based so the `auto` return type will always error. I've added a different integer test case:

```hlsl
int64_t4 HooBoy() {
  return 4l.xxxx;
}
```

`err()` works, which is also sadly expected despite the terrible syntax:
[FXC on shader playground](https://shader-playground.timjones.io/9111b2b533b5665e679c777fe4e5123b)

AaronBallman replied:
Please add a codegen test for `err()`

cor3ntin commented on clang/lib/Lex/LiteralSupport.cpp:
```diff
@@ -930,7 +930,11 @@ NumericLiteralParser::NumericLiteralParser(StringRef TokSpelling,
+  // 'pp-number' regex. This is required to support vector swizzles on numeric
+  // constants (i.e. 1.xx or 1.5f.rrr).
+  if (isPreprocessingNumberBody(*ThisTokEnd) &&
+      !(LangOpts.HLSL && *ThisTokEnd == '.')) {
```
This looks fine to me - if a bit hackish.

What surprises me is that the `if` existed at all before the change.

`LexNumericConstant` would have consumed everything it could. There should *never* be a valid pp-number after.

We should be able to write something like
```cpp
assert(!isPreprocessingNumberBody(*ThisTokEnd) || (LangOpts.HLS && *ThisTokEnd == '.'))
```

and remove that `if`  / diagnostic (which is never tested)

cor3ntin commented:
This is the documentation i found.
Can you confirm the intent is only to support `.x???` and `r???` ?
It alludes to more options.
Maybe we need a `isHLSLSwizzleStart` function to avoid comparing to `'x'` and `'r'` in multiple places.

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx9-graphics-reference-asm-vs-registers-modifiers-source-swizzling
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx9-graphics-reference-asm-ps-registers-modifiers-source-register-swizzling  

I don't suppose there is a specification / grammar for these things?

llvm-beanz commented:
> This is the documentation i found. Can you confirm the intent is only to support `.x???` and `r???` ? It alludes to more options. Maybe we need a `isHLSLSwizzleStart` function to avoid comparing to `'x'` and `'r'` in multiple places.

The documentation for this is bad. Single element vectors can only have `x` and `r` components, so it should never be anything else. Adding a helper is probably worth doing just for code simplicity and expressivity. We also don't support the `r` suffix, so there shouldn't be any syntactic ambiguity with things like `1.r` being interpreted as fixed point.

> https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx9-graphics-reference-asm-vs-registers-modifiers-source-swizzling https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx9-graphics-reference-asm-ps-registers-modifiers-source-register-swizzling

There is also this page:
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-per-component-math

It doesn't really add much.

> I don't suppose there is a specification / grammar for these things?

Working on it. I've been working to draft a language specification [over here](https://github.com/microsoft/hlsl-specs/tree/main/specs/language). I haven't yet gotten to the vector grammar. I'm unfortunately chasing down too many paths at once trying to both get enough syntax supported that we can generate real programs to flesh out codegen, and trying to get the language semantics documented and implemented in a way that conforms with the reference compilers.

I'm digging myself out from the Dev Meeting, but I'll work on getting an update to this PR posted this week.

Thanks for the great feedback!

llvm-beanz committed:
Updates bassed on review feedback.

../clang/test/SemaHLSL/Types/BuiltinVector/ScalarSwizzleErrors.hlsl

llvm-beanz committed:
[NFC] Fix clang-format violation

AaronBallman commented on clang/test/SemaHLSL/Types/BuiltinVector/ScalarSwizzles.hlsl:
```diff
@@ -0,0 +1,78 @@
+// CHECK-NEXT: DeclRefExpr {{.*}} 'float' lvalue ParmVar {{.*}} 'V' 'float'
+
+
+float4 ToThreeFloats(float V){
```
Isn't this to four floats?

llvm-beanz committed:
Codegen support

This change adds CodeGen support for writing to scalar l-values through
ext vector component l-value expressions. This is something only HLSL is
crazy enough to support.

../clang/test/CodeGenHLSL/builtins/ScalarSwizzles.hlsl
../clang/test/SemaHLSL/Types/BuiltinVector/ScalarSwizzleErrors.hlsl
../clang/test/SemaHLSL/Types/BuiltinVector/ScalarSwizzles.hlsl

llvm-beanz commented:
Friendly ping @AaronBallman & @cor3ntin.

llvm-beanz commented:
Friendly ping @AaronBallman & @cor3ntin.

cor3ntin reviewed (approved):
I am happy with the lexing/sema changes.
Please give @AaronBallman the opportunity to look at the code gen bits.

AaronBallman reviewed (approved):
LGTM!
